src/EditorFeatures/Core.Wpf/InlineRename/InlineRenameSession.cs

```diff
@@ -343,9 +347,18 @@ public void RefreshRenameSessionWithOptionsChanged(Option<bool> renameOption, bo
 
                 var cancellationToken = _cancellationTokenSource.Token;
 
-                UpdateReferenceLocationsTask(_allRenameLocationsTask.SafeContinueWithFromAsync(
-                    t => _renameInfo.FindRenameLocationsAsync(_optionSet, cancellationToken),
-                    cancellationToken, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default));
+                UpdateReferenceLocationsTask(ThreadingContext.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    // Join prior work before proceeding, since it performs a required state update.
+                    // https://github.com/dotnet/roslyn/pull/34254#discussion_r267024593
+                    //
+                    // The cancellation token is passed to the prior work when it starts, not when it's joined. This is
+                    // the equivalent of TaskContinuationOptions.LazyCancellation.
+                    await _allRenameLocationsTask.JoinAsync(CancellationToken.None).ConfigureAwait(false);
```

Comment:
Why wouldn't this just pass the cancellationToken? I think you're trying to preserve behavior, but do we need to?

(I never liked that LazyCancellation was forced on everybody using SafeContinueWith who didn't perhaps need it because it means we would be potentially cancelling things a bit later.)

Reply:
A behavior change here requires proof that the newly-allowed concurrency would not result in state corruption.

Reply:
FindRenameLocationsAsync already can be running multiple times in parallel if for example options were changed in rapid succession multiple times. Either it's already safe for concurrency (and it's not a concern) or it's not, and then this is still bugged.

Reply:
I was not able to prove that this one would behave correctly with lazy cancellation. With that said, this cancellation token only gets cancelled when the rename session is cancelled or committed, so it won't be impacting typing performance.

Reply:
> With that said, this cancellation token only gets cancelled when the rename session is cancelled or committed, so it won't be impacting typing performance.

Ah, maybe this is why I was confused earlier: this code has more problems than what you're trying to fix. :smile: If the options are changed, we should have entirely cancelled off the old work because there's no reason for that to continue running. What this *should* be doing (loosely) is cancelling _cancellationTokenSource and creating a new one before it calls UpdateReferenceLocationsTask. It appears this predates the open sourcing of this so I don't know why this ended up this way.

Want to file a bug? It's not that hard to fix but requires some care and I don't blame you for not wanting to fix it in this PR. (Although you're welcome to!)

Reply:
Filed #40891